File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
  public ListNode MergeTwoLists(ListNode list1, ListNode list2) => (list1, list2) switch {
    (null, null) => null,
    var (l1, l2) when l1 == null => l2,
    var (l1, l2) when l2 == null => l1,
    var (l1, l2) when (l1.val < l2.val) => new ListNode(l1.val, MergeTwoLists(l1.next, l2)),
    var (l1, l2) when (l1.val >= l2.val) => new ListNode(l2.val, MergeTwoLists(l1, l2.next))
  };
}

File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
  def mergeTwoLists(list1: ListNode, list2: ListNode): ListNode = {
    val dummy = new ListNode()
    var p1 = list1
    var p2 = list2
    var p = dummy
    while (p1 != null && p2 != null) {
      if (p1.x < p2.x) {
        p.next = p1
        p1 = p1.next
      } else {
        p.next = p2
        p2 = p2.next
      }
      p = p.next
    }
    if (p1 == null) {
      p.next = p2
    }
    if (p2 == null) {
      p.next = p1
    }
    return dummy.next
  }
}